### ISSUE TITLE
Support running benchmarks

### ISSUE THREAD

**Issue by mvtec-bergdolll** [Kobzol/cargo-pgo#5]
Currently I was only able to build an optimized benchmark binary with `cargo pgo --bench <bench>`. But I could find no way to feed it into the benchmark runner. I suspect something like `cargo pgo bench` and `cargo pgo bolt bench` would be required to support this. The only workaround I found was to copy the bolt optimized binary to the path of the regular `cargo bench` generated one, essentially replacing the binary. But that's rather cumbersome.

**Comment by Kobzol** [comment]
Supporting `cargo bench` is definitely on my TODO list, I just didn't do it yet since benchmarks are not stable. I will try to implement it soon.

**Comment by Kobzol** [comment]
I implemented support for gathering PGO profiles from benchmarks in https://github.com/Kobzol/cargo-pgo/pull/9. For BOLT, it's a bit more complicated and I'll have to implement it in another way, to also support running `test` and other things.

Could you please try if `cargo pgo bench` works for you with the following version of the crate?
```bash
$ cargo install --git https://github.com/kobzol/cargo-pgo --branch pgo-bench
```

**Comment by Kobzol** [comment]
I made two changes in https://github.com/Kobzol/cargo-pgo/pull/15:
- PGO profiles are now named after the hash of their content. This means that if you do this:
    1) Gather profile
    2) Optimize
    3) Gather more profiles
    4) Optimize
    
     The code will be recompiled properly. Before it could be stale, although it probably didn't occur in practice, since gathering profiles always clears the profile directory (I will probably change this default behaviour though).
- Merged profiles are not used for creating new merged profiles. This was a bug that would cause a forced recompilation on each `cargo pgo optimize` execution with the previously described change.

Even though it's not a 100 % fix for your use-case, this now means that if you do not change the code, and you do not add/remove/change profiles, running `cargo pgo optimize bench` two times in a row should not cause a recompilation, so you can use it to run the optimized benchmarks multiple times.

**Comment by pkolaczk** [comment]
I'm having trouble using this feature. I run `cargo pgo optimize bench` and `top`confirms my benchmark is running, however I can see no benchmark output:

```
$ cargo pgo optimize bench
[2022-10-03T07:05:01Z INFO  cargo_pgo::pgo::optimize] Found 2 PGO profile files with total size 2.21 MiB at /home/pkolaczk/Projekty/DataStax/shadow/target/pgo-profiles.
[2022-10-03T07:05:01Z INFO  cargo_pgo::pgo::optimize] Merged PGO profile(s) to /home/pkolaczk/Projekty/DataStax/shadow/target/pgo-profiles/merged-606e657a2ce0ebae4f35e80ea999c3425e9b26dc1c0a8035252f2594bf11652e.profdata.
```

I'm using criterion v0.4 for benchmarking.

Does pgo bench do any non-standard output capturing / buffering?

**Comment by pkolaczk** [comment]
Ah, ok, so it looks like it does capture stderr/stdout and print it out only after all benchmarks are finished. 
Would be nice if it could actually print them out as it goes, because it is a long running process.

**Comment by Kobzol** [comment]
I created a new issue for this: https://github.com/Kobzol/cargo-pgo/issues/19. It will require a bit of refactoring of the plugin, I will try to fix this later this week.

**Comment by chadbrewbaker** [comment]
* Quick and dirty

Dumping /usr/bin/time -v counters. (/usr/bin/time -l for BSDTIME)

* Function specialization and type shrinking

Run real world integration test data with https://github.com/namhyung/uftrace to collect histograms of each variable and lint use of types that are too large. i.e. using a uint64 when a small enum is enough. 

https://github.com/oracle/graal/tree/master/sulong to run the LLVM code under a JIT and dump the JIT function specializations as lints.

* Detect Thread bottlenecks

Run https://github.com/plasma-umass/coz to detect threads that are bottlenecks.

* Worst case performance 

https://github.com/klee/klee and fuzzers

* Superoptimization of expensive code

Take most expensive code and blast with souper https://github.com/google/souper

* Safety for free

Run integration tests under LLVM sanitizers, MIRI, valgrind.

**Comment by Kobzol** [comment]
Both running and optimizing benchmarks should now be supported, and their output is streamed. Led me know if there's more that I can do with `cargo pgo`.
